### PR TITLE
ARTEMIS-3660 - rename broker balancer to connection router - tidy up config

### PR DESCRIPTION
**Pull request by gtully**
…config attributes in broker balancer

### CODE REVIEW

**Comment by gtully** [comment]
I wonder should I go further and replace the name brokerBalancer with "distributor"?

in that way, an acceptor would redirectTo a distributor rather than a balancer. The notion of balancing is gone, as that is really a feature or not of the distribution choices or policy

**Comment by brusdev** [comment]
@gtully I agree with you: `target` in the configuration is a little confusing. What about `ConnectionKey` to replace `TargetKey`and `ConnectionHandler` to replace `BrokerBalancer`. So acceptor can define the connectionHandler that will handles the connections and the ConnectionHandler can define the key of the connection to apply its filters/polices.

**Comment by gtully** [comment]
ah, your comments are relating to renaming the classes... yes. it would make sense to change those names. I am primarily concerned with the configuration at the moment. as that is the user facing bit.

**Comment by jbertram** [comment]
Conflicts need to be addressed before this can be merged.

**Comment by gtully** [comment]
rebased!

**Comment by gtully** [comment]
@brusdev I am coming around to your idea of connection-handler, distributor is possibly too descriptive and restrictive. It could be a distributor or partitioner or balancer - it depends. As a connection-handler it remains up to the configuration as to what it actually means. The name attribute is the ultimate place for the pithy description.

**Comment by gtully** [comment]
@brusdev new proposal, in place of broker balancer - use connection router:
````
<connection-routers><connection-router name="partition_by_client_id" .. /><connection-routers>  
<acceptor .... router="partition_by_client_id" />
````

**Comment by brusdev** [comment]
@gtully connection-router is perfect +1000

**Comment by gtully** [comment]
done the rename, balancer -> routing package, Balancer to Router in class names to reflect the xml config

**Comment by pwright** [comment]
@gtully should an update to https://github.com/apache/activemq-artemis/blob/main/docs/user-manual/en/broker-balancers.md be part of this PR?

**Comment by gtully** [comment]
@pwright for sure and there is: https://github.com/apache/activemq-artemis/pull/3932/files?file-filters%5B%5D=.md&show-viewed-files=true

**Comment by pwright** [comment]
@gtully doh! I searched the diff for .md, but TIL how to filter files in a PR 😉

**Comment by brusdev** [comment]
@gtully I'll push a few commits to finish refactoring for doc, packaging, ..., feel free to drop or squash them

**Comment by gtully** [comment]
some more internal class rename to follow form @brusdev - leaving as draft till complete. Best to get this all done at once.

**Comment by brusdev** [comment]
@gtully could you review the following POC to collapse KeyTransformer and Policy, to reduce routing components?
https://github.com/brusdev/activemq-artemis/commit/adea269bb65fe0c71dcf55bf242a6f0f2e9a8729

**Review by brusdev — APPROVED** [review]

File: examples/features/connection-router/evenly-redirect/src/main/resources/activemq/server0/broker.xml [hunk 1]
```diff
@@ -46,18 +46,18 @@ under the License.
 
       <acceptors>
          <!-- Acceptor for every supported protocol -->
-         <acceptor name="artemis">tcp://0.0.0.0:61616?redirect-to=evenly-balancer;tcpSendBufferSize=1048576;tcpReceiveBufferSize=1048576;amqpMinLargeMessageSize=102400;protocols=CORE,AMQP,STOMP,HORNETQ,MQTT,OPENWIRE;useEpoll=true;amqpCredits=1000;amqpLowCredits=300;amqpDuplicateDetection=true</acceptor>
+         <acceptor name="artemis">tcp://0.0.0.0:61616?router=evenly-router;tcpSendBufferSize=1048576;tcpReceiveBufferSize=1048576;amqpMinLargeMessageSize=102400;protocols=CORE,AMQP,STOMP,HORNETQ,MQTT,OPENWIRE;useEpoll=true;amqpCredits=1000;amqpLowCredits=300;amqpDuplicateDetection=true</acceptor>
```
Comment (gtully) [inline examples/features/connection-router/evenly-redirect/src/main/resources/activemq/server0/broker.xml:49]: not a biggie, but this router could be called 'evenly-balance'

File: examples/features/connection-router/symmetric-simple/readme.md [hunk 2]
```diff
@@ -24,8 +24,8 @@ In configuration, the `local-target-filter` provides the regular expression that
 `Broker0` takes clientIDs with prefix FOO and `Broker1` takes prefix BAR. The `target-key-filter` specifies how the key is extracted,
```
Comment (gtully) [inline examples/features/connection-router/symmetric-simple/readme.md:24]: there is a target-key prefix that has slipped through :-)

Comment (gtully) [inline examples/features/connection-router/symmetric-simple/readme.md:24]: there is a 'target-' prefix that has slipped through I think :-)

**Pull request merged**